sklearn/metrics/_pairwise_distances_reduction/_argkminlabels.pyx

```diff
@@ -0,0 +1,203 @@
+        """
+        pda = PairwiseDistancesArgKminLabels64(
+            datasets_pair=DatasetsPair.get_for(X, Y, metric, metric_kwargs),
+            k=k,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            weights=weights,
+            labels=labels,
+        )
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if pda.execute_in_parallel_on_Y:
+                pda._parallel_on_Y()
+            else:
+                pda._parallel_on_X()
+
+        return pda._finalize_results(return_distance)
+
+    def __init__(
+        self,
+        DatasetsPair datasets_pair,
+        chunk_size=None,
+        strategy=None,
+        ITYPE_t k=1,
+        weights=None,
+        cnp.ndarray[ndim=1, dtype=ITYPE_t] labels=None,
+    ):
+        super().__init__(
+            datasets_pair=datasets_pair,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            k=k,
+        )
+
+        if weights == "uniform":
+            self.weight_type = Weight.uniform
+        elif weights == "distance":
+            self.weight_type = Weight.distance
+        else:
+            self.weight_type = Weight.other
+        self.labels = labels
+        self.unique_labels = np.unique(labels)
+
+        # Heap to be returned
+        self.argkmin_labels = np.full(self.n_samples_X, -1, dtype=ITYPE)
+
+        # Map from set of unique labels to their indices in `label_counts`
+        self.labels_to_index = {label:idx for idx, label in enumerate(self.unique_labels)}
+        # Buffer used in building a histogram for one-pass weighted mode
+        self.label_counts = np.zeros((self.n_samples_X,  len(self.unique_labels)), dtype=DTYPE)
+
+    def _finalize_results(self, bint return_distance=False):
+        if return_distance:
+            # We need to recompute distances because we relied on
+            # surrogate distances for the reduction.
+            self.compute_exact_distances()
+
+            # Values are returned identically to the way `KNeighborsMixin.kneighbors`
+            # returns values. This is counter-intuitive but this allows not using
+            # complex adaptations where `PairwiseDistancesArgKmin64.compute` is called.
+            return np.asarray(self.argkmin_distances), np.asarray(self.argkmin_indices), np.asarray(self.argkmin_labels)
+
+        return np.asarray(self.argkmin_indices), np.asarray(self.argkmin_labels)
+
+    cdef void _parallel_on_X_prange_iter_finalize(
+        self,
+        ITYPE_t thread_num,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+    ) nogil:
+        cdef:
+            ITYPE_t idx, jdx, y_idx, label_index, sample_index
+            DTYPE_t max_weight
+            ITYPE_t label, max_label
+            DTYPE_t val, weight
```

Comment:
Not sure of the semantic of `class` vs. `label`, especially regarding multi-class and multi-output (cc @thomasjpfan).

Let's keep `label` but precise the variables' nature.
```suggestion
            ITYPE_t idx, jdx, y_idx, label_index, sample_index
            ITYPE_t label, max_label
            DTYPE_t max_label_weight,
            DTYPE_t label_weight, nn_weight
```